// angular13-organization-chart/esm2020/src/app/modules/org-chart/org-chart.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { OrgChartComponent } from './org-chart/org-chart.component';
import * as i0 from "@angular/core";
export class OrgChartModule {
}
OrgChartModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.6", ngImport: i0, type: OrgChartModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
OrgChartModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.6", ngImport: i0, type: OrgChartModule, declarations: [OrgChartComponent], imports: [CommonModule], exports: [OrgChartComponent] });
OrgChartModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.6", ngImport: i0, type: OrgChartModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.6", ngImport: i0, type: OrgChartModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule
                    ],
                    declarations: [OrgChartComponent],
                    exports: [
                        OrgChartComponent
                    ]
                }]
        }] });
