// ngx-filter-pipe/esm2022/src/app/shared/ngx-filter-pipe/ngx-filter.module.mjs
/**
 * Created by vadimdez on 28/11/2016.
 */
import { NgModule } from '@angular/core';
import { FilterPipe } from './ngx-filter.pipe';
import * as i0 from "@angular/core";
export class FilterPipeModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.8", ngImport: i0, type: FilterPipeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.8", ngImport: i0, type: FilterPipeModule, declarations: [FilterPipe], exports: [FilterPipe] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.8", ngImport: i0, type: FilterPipeModule, providers: [FilterPipe] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.8", ngImport: i0, type: FilterPipeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [FilterPipe],
                    providers: [FilterPipe],
                    exports: [FilterPipe]
                }]
        }] });
